sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,314 @@
+
+def disallow_response(bot, nick: str, email):
+    fdrop(bot, nick)
+    add_badmail(bot, email)
+    bot.say(f"Your registration has been disallowed because {email.domain} appears to be suspicious. "
+             "If this is a legitimate domain, contact staff for assistance.",
+             nick.lower())
+    alert(bot, f"WARNING: User {nick} attempted to register a nick with suspicious domain {email.domain}.")
+
+def fetch_IPQS_email_score(
+    email_addr: str,
+    key: str,
+    fast: bool = True
+    ) -> tuple[float, bool, bool]: #score, disposable, has recent abuse flag set
+    '''Perform lookup on a specific email adress using ipqualityscore.com'''
+    email_str = urllib.parse.quote(email_addr)
+    faststr = str(bool(fast)).lower() #lower + handle None and other garbage
+    params = urllib.parse.urlencode({'fast': faststr})
+    with urllib.request.urlopen(
+        f"https://ipqualityscore.com/api/json/email/{key}/{email_str}?{params}") as url:
+            data = json.loads(url.read().decode())
+            LOGGER.debug(data)
+    if not data['success']:
+        errstr = f"{email_addr} lookup failed with {data['message']}"
+        LOGGER.error(errstr)
+        raise RuntimeError(errstr)
+    return (data['fraud_score'], data["disposable"], data["recent_abuse"])
+
+def get_email_score_from_db(session, email):
+    query_result = session.query(KnownEmails)\
+        .filter(KnownEmails.domain == email.domain)\
+        .one_or_none()
+    if query_result:
+        #Any known problematic provider should've been BADMAILed by now, but...
+        return DomainInfo(query_result.score,
+                          query_result.flag_disposable,
+                          query_result.flag_recent_abuse)
+
+def store_email_score_in_db(session, email, nick, IPQSresult):
+    new_known_email = KnownEmails(domain= email.doman[:DOMAIN_LEN],
+                                    first_nick= nick,
+                                    score= IPQSresult[0],
+                                    flag_disposable= IPQSresult[1],
+                                    flag_recent_abuse= IPQSresult[2])
+    session.add(new_known_email)
+    session.commit()
+
+def retrieve_score(bot, email, nick):
+    session = bot.db.session()
+    try:
+        global db_populated
+        if not db_populated:
+            BASE.metadata.create_all(bot.db.engine)
+            db_populated = True
+        if retval := get_email_score_from_db(session, email):
+            return retval
+        else:
+            if IPQSresult := fetch_IPQS_email_score(email, bot.config.emailcheck.IPQS_key):
+                store_email_score_in_db(session, email, nick, IPQSresult)
+                return IPQSresult
+            else: #Shouldn't be possible
+                raise RuntimeError(f"Couldn't retrieve IPQS for {email}!")
+    except SQLAlchemyError:
+        session.rollback()
+        raise
```

Comment:
Since you have more than one function doing DB calls, I would suggest a wrapper that lets you do `with ...` syntax for transactions that automatically commits and rolls back.